### ISSUE TITLE
operator== documentation should show how to apply custom comparison function

### ISSUE THREAD

**Issue by nmusolino** [nlohmann/json#1915]
## What is the issue you have?

The [documentation for `operator ==`](https://nlohmann.github.io/json/classnlohmann_1_1basic__json_a122640e7e2db1814fc7bbb3c122ec76e.html) mentions that double NaN values compare unequal, and says, "To compare floating-point while respecting an epsilon, an alternative comparison function could be used..."  The example function is for comparing two doubles.

In issue #514, about comparison between NaN values, @nlohmann wrote:

> I think nothing must be changed here, only the documentation must be clear.

The documentation for `operator ==` should illustrate *how* a custom double comparison function could be applied recursively to all values in a `basic_json` object.  This would allow users to easily write their own function to evaluate two `basic_json` objects for equivalence.

I think such a function could be written using `basic_json::flatten()` and comparing items pairwise with the std::equal algorithm.

**Comment by nlohmann** [comment]
What I meant with "an alternative function" would be a user-defined function like

```cpp
bool my_equal(const json& lhs, const json& rhs)
{
    switch (lhs.type())
    ...
}
```

which implements whatever logic you would like. There is no extension point to change the behavior of `operator==`.

**Comment by nmusolino** [comment]
I understand.

What I am trying to suggest is this... let's say that a user has a custom comparison in mind, like the example in [the documentation page](https://nlohmann.github.io/json/classnlohmann_1_1basic__json_a122640e7e2db1814fc7bbb3c122ec76e.html), or like this:
```c++
inline bool is_same(double a, double b) noexcept
{
    return a == b || (std::isnan(a) && std::isnan(b)); 
}
```

Starting from that point, a user would like to evaluate whether two arbitrarily nested json objects are "equivalent."  My suggestion is that the documentation include a working version of this function:

```c++
bool my_equal(const json& lhs, const json& rhs)
{
    // Uses `my_equal` to compare doubles.
    switch (lhs.type())
    ...
}
```

By the way, I think this would be facilitated if json had a `visit` helper function as requested in #1458.  Users could write a binary visit function on top of that.

**Comment by dota17** [comment]
Comparing floating point with `==` directly is unsafe and dangerous, and it seems that compare their absolute value with epsilon is not perfect too, according to [THE FLOATING-POINT GUIDE](https://floating-point-gui.de/errors/comparison/). @nmusolino @nlohmann Should we discuss how to deal with this situation?

**Comment by nlohmann** [comment]
- Regarding the visit helper function, please open an issue.
- Changing the behavior of `operator==` would be a breaking change that I would like to avoid.
- I agree that putting such examples into a dedicated places is a good idea, see #1953.

**Comment by Stannieman** [comment]
I also think we should by default use the default operator== for all types.
Especially for floating point numbers there is no one size fits all way of comparing them.
Whether we should use normal ==, with epsilon or some more complex mechanism depends totally on the kind of data contained in the JSON.
Therefore, in my opinion, allowing the user to provide a custom comparison is the only correct option here.

**Comment by nlohmann** [comment]
Closed by merging #1984.
